### ISSUE TITLE
Mark payments as refunded with webhook listener (Stripe)

### ISSUE THREAD

**Issue by thomasfw** [restrictcontentpro/restrict-content-pro#1413]
The Stripe webhook listener doesn't currently process the **charge.refunded** event. It would be useful if payments refunded in Stripe were automatically marked as such in RCP - should be pretty simple to achieve with an extra `$event->type` conditional in the gateway's `process_webhooks()`.

**Comment by mindctrl** [comment]
We should probably make sure we do this for all the gateways.

Also probably a good idea to mark the member as expired when the refund happens.

**Comment by thomasfw** [comment]
Just a reminder not to automatically expire the member on the refund event until #962 is resolved

**Comment by nosegraze** [comment]
PR: #1540

This works for Stripe now. Going to check the other gateways.

**Comment by nosegraze** [comment]
- [x] 2Checkout (already done)
- [ ] Authorize.net
- [ ] Braintree
- [ ] PayPal Standard
- [ ] PayPal Express/Pro
- [x] Stripe / Stripe Checkout (done in PR)

**Comment by nosegraze** [comment]
PR is ready now. The only one I haven't tested yet is PayPal Standard. Sandbox was acting up this morning.

Also see note about Braintree... looks like they don't send webhooks for refunds issued inside the dashboard.

**Comment by mindctrl** [comment]
I see `// @todo Update member status to expired.` in these.

I see [this comment](https://github.com/restrictcontentpro/restrict-content-pro/issues/1413#issuecomment-332031121) mentions not marking the member expired until #962 is resolved, but that's unrelated to refunds unless I'm not thinking about something (quite possible). Can't we go ahead and add the member expiration bits?

**Comment by thomasfw** [comment]
@mindctrl we're having to go in and refund duplicate payments and cancel the duplicated subscriptions caused by #962. Those members are still active (they've just renewed in fact) and should not be expired, so adding the auto-expiry into the webhook listener would cause more issues until that issue (now #1570) is fixed.

**Comment by nosegraze** [comment]
^ That's exactly why I omitted it. The need to refund duplicate payments is still there until we fix https://github.com/restrictcontentpro/restrict-content-pro/issues/1570 .

**Comment by mindctrl** [comment]
Right, but can't we determine if the refunded payment is associated with their current subscription or not, and only expire the member if it is? We don't want to always expire on refunds, we need to check that the payment is related. Related: https://github.com/restrictcontentpro/restrict-content-pro/issues/1503

**Comment by nosegraze** [comment]
Well the payments are probably also getting logged in RCP as well, so RCP would think they were associated. Not 100% sure though.

**Comment by tnorthcutt** [comment]
One note on changing status to expired after a refund: in Stripe, at least, you can issue a partial refund.  I have a client who has to do this sometimes due to https://github.com/restrictcontentpro/restrict-content-pro/issues/1700. If a refund in Stripe automatically results in a member's status changing to expired, that would be... problematic.

Another non-bug-dependent use-case would be refunding a month's payment for customer service reasons, promotional reasons, etc. Basically, I don't think automatically expiring a member as a result of a refund is a safe action.

**Comment by mindctrl** [comment]
On partial refunds, we can check the refunded amount and if it's not a full refund, adjust the payment amount in RCP and leave its status as is.

**Comment by tnorthcutt** [comment]
Sounds good. For reference, in Stripe, the `amount_refunded` property on the `charge.refunded` event object holds the amount refunded (in cents). /cc @nosegraze
